pkg/endpoint/bpf.go

```diff
@@ -670,6 +688,14 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir, reason string) (uint64, err
 	rundir := owner.GetStateDir()
 	debug := strconv.FormatBool(owner.DebugEnabled())
 
+	// To avoid traffic loss, wait for the proxy to be ready to accept traffic
+	// on redirect ports, before we generate the policy that will redirect
+	// traffic to those ports.
+	err = e.WaitForProxyCompletions()
```

Comment:
in an offline discussion with @raybejjani in came to this thoughts:
We need this change because the policy goes from "no policy" to "l7 enforcement" and we need to get the proxy ready before sending the traffic from bpf to the proxy.

The question now is: isn't also problem when we go from "l7 enforcement" to "no policy"?

Reply:
Good point. So I should move the `e.removeCollectedRedirects(owner)` call until after the call to `e.runInit(...)`. I will need to create and use another WaitGroup and call Wait a 2nd time for that call, since we want to exclude the `e.runInit(...)` call from any WaitGroup wait / context timeout. Something like:
```
	completionCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	e.ProxyWaitGroup = completion.NewWaitGroup(completionCtx)
	defer func() {
		cancel()
		e.ProxyWaitGroup = nil
	}()
	...
	e.allocateRedirects(...)
	...
	err = e.WaitForProxyCompletions()
	if err != nil {
		return 0, fmt.Errorf("Error while configuring proxy redirects: %s", err)
	}

	err = e.runInit(...)

	completionCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	e.ProxyWaitGroup = completion.NewWaitGroup(completionCtx)
	defer func() {
		cancel()
		e.ProxyWaitGroup = nil
	}()

	e.removeCollectedRedirects(owner)

	err = e.WaitForProxyCompletions()
	if err != nil {
		return 0, fmt.Errorf("Error while configuring proxy redirects: %s", err)
	}
```

@aanm, @raybejjani, does that make sense to you?

Reply:
@rlenglet When making this change should check that `proxiesToRemove` does not contain any of the proxies we just added. I don't now recall how we name proxies, but should check that it is not possible to delete a proxy and then add the same proxy again, and have the implementation result in the state where the proxy is first added and then deleted, having no proxy at the end.